mlir/lib/Dialect/MIOpen/Generator/IsaNameParser.cpp

```diff
@@ -0,0 +1,128 @@
+//===- IsaNameParser.cpp - MLIR to C++ option parsing ---------------===//
+//
+// Part of the MLIR Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements isa name string parser
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/MIOpen/Generator/IsaNameParser.h"
+#include "llvm/Support/Error.h"
+
+#include <cstring>
+#include <numeric>
+#include <vector>
+
+using namespace mlir;
+
+static constexpr const char gcnArchDelimiter[] = ":";
+
+namespace {
+LogicalResult getTripleFromIsaName(const std::string &isaName,
+                                   std::string &triple) {
+  std::size_t firstSeperatorLoc = isaName.find(gcnArchDelimiter);
+  if (firstSeperatorLoc == std::string::npos) {
+    return failure();
+  }
+  triple = isaName.substr(0, firstSeperatorLoc);
+  return success();
+}
+
+std::string getChipFromArchName(const std::string &gcnArchName) {
+  std::size_t firstSeperatorLoc = gcnArchName.find(gcnArchDelimiter);
+  if (firstSeperatorLoc == std::string::npos) {
+    return gcnArchName;
+  }
+
+  return gcnArchName.substr(0, firstSeperatorLoc);
+}
+
+// This function converts the arch name to target features:
+// sramecc+:xnack- to +sramecc,-xnack
+LogicalResult parseTargetFeatures(std::string &gcnArchFeatures) {
+  LogicalResult status = success();
+  // First step: put each feature name to the vector
+  std::string token;
+  std::vector<std::string> featureTokens;
+  auto convertFeatureToken = [&status](std::string &token) {
```

Comment:
That definitely seems to be a more robust way to handle it. Though for consistency, we have mostly used LogicalResult rather than Error to capture error cases. I'd probably make this lambda return a LogicalResult instead to make sure failure is always caught.

Reply:
Well, the reason I'm suggesting Expected<> is that we want to return either a string or failure, as opposed to maybe returning a valid string and signalling failure. Having that union type in the signature seems like a better thing to do and makes the code clearer and less error-prone (IMO, but then again, I'm coming from a Rust/Haskell/... background)

Reply:
> want to return either a string or failure, as opposed to maybe returning a valid string and signalling failure

This is very convincing in this lambda. 

I will address this feedback but I think this can potentially apply to lots of places we are currently using `LogicalResult` too. For the scope of this review I will only amend this single occurrence but I will try to gain more conscious of similar cases in the future.

Reply:
Thanks - I know that these sort of types (`Expected<>`, `Optional<>`, and so on) aren't native to C++ and tend to be bolted on, and so it makes sense why we don't see them too often in our codebase. My instinct towards them is definitely a function of my background (and given who the LLVM people are, I can see why there's a lot of that sort of thing in the LLVM headers).

Reply:
Done in 4981d203a6c728f3f0698b2eb2bd1b0bec86a190.

I did a little investigation of how error handling has been done around llvm/mlir. Looks like we are making the right decisions here:
 - In ExecutionEngine, errors are done by Expected<T>
 - In Dialect library, errors are done by either `asserts` or `emitError` (which gives location info to wrong mlir code)